### ISSUE TITLE
[23] super reference disallowed in the context of a field access

### ISSUE THREAD

**Issue by jarthana** [eclipse-jdt/eclipse.jdt.core#3094]
Here's the testcase:
```
public class X {
	class Nested extends X1 {
		X1 xy;
		class DeeplyNested extends NestedInX1 {
			DeeplyNested(float f) {
				Nested.super.x1.super(); // Error here
			}
		}
	}
}
class X1 {
	X1 x1;
	class NestedInX1 {}
}

```
The error reported is `Cannot use 'Nested.super' in an early construction context (except in a simple field assignment)`
I don't really know the intricacies here and Javac allows this code.

Copying @mpalat and @stephan-herrmann

**Comment by jarthana** [comment]
Here's another variant:
```
    Y(int a) {
        X.field = new Y() {
            public int foo() {
                return this.a;
            }
        }.foo();
        this();
    }
```

**Comment by jarthana** [comment]
> @jarthana Which _field_ should `this.a` point to? Somehow you're not telling the full story of that snippet.

Yeah, I was just playing around and was too eager to report. Let me get back on this.

**Comment by stephan-herrmann** [comment]
> The error reported is `Cannot use 'Nested.super' in an early construction context (except in a simple field assignment)` 

The naked eye doesn't even see any early construction context (and --enable-preview is not needed for javac). So this part looks buggy ...

> I don't really know the intricacies here and Javac allows this code.

"intricacies" is a good word here :)
It took my quite some time to figure out what this program is trying to say.

This version seems to be equivalent, and is accepted by ecj:
```
	x1.super();
```
(qualifying the inherited field is not necessary).

Read: when going from DeeplyNested to its super NestedInX1 we need one additional enclosing instance which is provided as `x1`. Wow!

**Comment by stephan-herrmann** [comment]
> public class X {
> 	class Nested {
> 		Nested(Object o) {}
> 	}
> 	class AnotherNested extends Nested {
> 		AnotherNested() {
> 			super(new Object() { // Cannot instantiate class new Object(){} in an early construction context of class X.AnotherNested
> 			});
> 		}
> 	}
> }

For this I filed #3132
